### ISSUE TITLE
Get Events for Tx event of type `txBestBlocksState`

### ISSUE THREAD

**Issue by Tbaut** [polkadot-api/polkadot-api#697]
Last one for today, as I continue going down the papi hole.

I've noticed one thing that I feel is a step back in terms of UX for my users, is that per default, they'd have to wait much much longer to see something change (say the balance update if they do a TX), because per default, everything is conservatively set to `finalized`, whereas they're used to have some feedback within 6s. This can be set for each query, but I'd argue `best` could be the default, as it's been like this in the past, and waiting more than 20s is simply too long.

Related to this, and more in line with the issue title, I'd like to know whether my TX has issues when it's in the best block. However, I could only get the events of a Tx when the `TxEvent` has the `event.type === finalized` which takes too long as stated before.

Now where you could help me is for the following:
I said I want the events, just because I'm actually digging them, to see if my Txs are truely ok.
I read in your docs that the `TxFinalized` type has a boolean "ok which simply tells if the extrinsic was successful (i.e. event System.ExtrinsicSuccess is found". However, with multisig transactions, or proxy or batch transactions, the tx can very well return a `System.ExtrinsicSuccess` although the actual execution of the call has failed, and to know this, there's no other way I know than to actually go check the results of these events.

The code I had to go deeper with pjs is horrendous, this is what pjs/apps has for instance, where incomplete message means something went wrong
```ts
export function getIncompleteMessage({ event }: EventRecord): string | null {
  const { method, section } = event

  return !!CHECKS[section] && !!CHECKS[section][method] && CHECKS[section][method](event)
}
```
The CHECKS can be found here, and this is not nice: https://github.com/polkadot-js/apps/blob/97aeb41a593165f9a9b218c1ceecb52172473de3/packages/react-components/src/Status/checks.ts

And then you check for incomplete message, and shows the 
 ```ts
       let errorInfo = ''
      const incomplete = getIncompleteMessage({ event } as EventRecord)

      // check if multisig or proxy or batch has an error
      if (incomplete) {
        errorInfo = incomplete
      }

      // if the extrinsic fails alltogether
      if (!errorInfo && api.events.system.ExtrinsicFailed.is(event)) {
        // extract the data for this event
        const dispatchError = data[0]
```

TL;DR I would like to show errors to my users when the proxy, multisig, or batch call they're doing errors out, without having to dig into event and I want to show them this asap, meaning when the tx is in the block, even when it's not finalized.

**Comment by Tbaut** [comment]
Thank you so much for the pointers, I totally missed the fact that I need to check the `found === true` in the `TxBestBlocksState` to get the events.

I see where you're coming from regarding the finalized data. That instructors at PBA didn't realize that there's absolutely no chance to have a UI reacting in 6s, while tracking a finalized block  is concerning though. They for sure didn't think about it and I hope they'd have come to the conclusion alone, had they known that Polkadot finality is 3 to 4 blocks behind.

The fact that pjs, and every other Dapp out there, have been using best block, and that I've personally never experienced nor heard any issue regarding this, makes me think this is totally fine, I may be wrong. Now on the UX front, _if the results are the same in most cases_, there's absolutely no question about the fact that having a 6s reaction time, is a much better experience. In any case it's great that everyone has the choice, and I understand that you prefer to remain conservative.

`signSubmitAndWatch` is what I use, and my quote comes from the docs.
> ok which simply tells if the extrinsic was successful (i.e. event System.ExtrinsicSuccess is found


I'll open a specific issue, definitely a not super important one, to suggest having some helper function so that Dapps don't have to know about all the calls, to show errors.

**Comment by josepot** [comment]
> I see where you're coming from regarding the finalized data. That instructors at PBA didn't realize that there's absolutely no chance to have a UI reacting in 6s, while tracking a finalized block is concerning though. They for sure didn't think about it and I hope they'd have come to the conclusion alone, had they known that Polkadot finality is 3 to 4 blocks behind.

I think that you are conflating 2 different things. What I’m pointing out is that when querying state without specifying a block, the default should logically be the finalized block. Either the `at` parameter is mandatory (as per the new JSON-RPC spec), or, if it's optional, the default block should be something that is not in a pending state.

> The fact that pjs, and every other Dapp out there, have been using best block, and that I've personally never experienced nor heard any issue regarding this, makes me think this is totally fine, I may be wrong.

In my experience, there *have* been issues caused by this approach, particularly in situations involving forks.

> Now on the UX front, _if the results are the same in most cases_,

That’s a big “if” and really depends on several factors like network congestion, client connectivity, number of peers, etc. In a consensus mechanism like BABE, which is more susceptible to forks than SAFROLE, forks occur frequently enough that relying on the best-block for state can lead to unpredictable behavior.

You can absolutely encounter a scenario where a transaction changes state on one fork but does something different on another fork. For example, a transaction could be front-run on one fork but not another, included in block x in one fork and in block x + 2 in another fork, or even be valid in one fork but invalid in another. The point is, the best-block is just the current view of the client, and that view can change.

> there's absolutely no question about the fact that having a 6s reaction time is a much better experience.

I agree that fast feedback is important for UX, and there is nothing preventing you from doing this with PAPI, but IMO DApps that offer this should present it as an optimistic update, clearly indicating to users which data is final and which data is "pending". Mixing best-block state with finalized state without distinction creates a misleading user experience and can lead to potential issues when forks happen.
